content/blog/2020-12-09-slp/index.md

```diff
@@ -0,0 +1,184 @@
+  where rather than have a single instruction execute, multiple instructions 
+  are executed each cycle. This is very similar to using SIMD instructions but 
+  requries no developer changes and is processor dependent. Techniques such as 
+  loop fission and loop unrolling can increase parallelism by reducing data
+  dependencies.
+
+## Superword Level Parallelism
+
+Superword level parallelism is the type of parallelism that the authors target. 
+This can be thought of as the SIMD subset of instruction level parallelism. It 
+is a sequence of short SIMD operations where the source and result data are 
+packed into a SIMD register. Additional operations to pack the initial data into 
+the source registers and some more operations to unpack the result might be
+needed.
+
+The authors devise a simple algorithm to convert a basic block that has
+opportunities for SLP into one that exploits SIMD registers and instructions.
+Broadly, the algorithm looks for sets of independent (no data dependencies),
+isomorphic (same operation) statements to merge together.
+
+### Algorithm
+
+The algorithm proceeds in 4 steps:
+
+1. First, identify pairs of independent, memory loads from adjacent locations.
+   These are natural candidates for SIMD loads. This is generally done by some
+   enabling analysis as an initial step. We add these adjacent loads to a set, 
+   effectively the set of lists of independent, isomorphic instructions that we 
+   want to combine into a single SIMD instruction.
+   ![phase1](phase1.png)
+2. We now need to extend this set to contain non-memory operations. Since we
+   want to take advantage of loaded data, we look for operations that use the 
+   packed data that we loaded before. We do this by following the uses and defs 
+   on the list of instructions that we packed. We can combine two statements 
+   into the current set:
+   - The statements do the same operation on independent data.
+   - It is beneficial to execute these at once rather than as multiple 
+     instructions. If the data is already packed, then there's no additional 
+     cost. However, if the data is not packed we have to take into account the 
+     cost of moving them into the SIMD register.
+   - The positions of the values in the SIMD register is consistent with the 
+     previous packed statements. This is what the paper calls alignment.
+   - To simplify the search space for the next part, we 
+     only allow a statement to occupy the beginning or end instruction in a 
+     pack list once. This way, we don't have to choose between multiple options 
+     later down the line.
+   ![phase2](phase2.png)
+3. Now that we have pairs of instructions that can be executed by a single SIMD
+   instruction without any issues, we can combine multiple of these pairs into 
+   an even wider SIMD instruction. Taking advantage of the last bullet above, 
+   we only combine two lists if the beginning and end instruction is the same.
```

Comment:
Not quite sure I got that last part—why do we want just the beginning and ending instructions to be the same? Wouldn't we want the instructions to be identical throughout?

Reply:
Clarified this.